[Bug] AOT codegen: ランタイムプリミティブが `user_*` にリネームされ gcc がコンパイル失敗する

## 概要 (Overview)

`cabal run spinor -- build <file>.spin` で AOT コンパイルを行うと、生成された C コード内でランタイム提供のプリミティブ (`print` など) が `user_*` プレフィクス付きの未定義シンボルとして出力されてしまい、後段の gcc がコンパイル失敗する。

`Spinor.Compiler.Codegen` の `mangle` 関数が、ユーザー定義シンボルとランタイムプリミティブを区別せず、すべてのシンボルに `user_` プレフィクスを付与しているのが原因と推定される。

## 再現手順 (Reproduction)

```bash
echo '(print "Hello, World!")' > hello.spin
cabal run spinor -- build hello.spin
```

### 実際の出力

```
Compiling hello.spin to hello.c...
Building hello with C:\msys64\ucrt64\bin\gcc.exe...
Build failed. C compiler output:

hello.c: In function 'main':
hello.c:8:14: error: implicit declaration of function 'user_print'; did you mean 'sp_print'? [-Wimplicit-function-declaration]
    8 |     sp_print(user_print(sp_make_str("Hello, World!")));
      |              ^~~~~~~~~~
      |              sp_print
hello.c:8:14: error: passing argument 1 of 'sp_print' makes pointer from integer without a cast [-Wint-conversion]
In file included from hello.c:3:
runtime/spinor.h:87:25: note: expected 'SpObject *' but argument is of type 'int'
   87 | void sp_print(SpObject* obj);
      |               ~~~~~~~~~~^~~
```

### 生成された hello.c

```c
#include <stdio.h>
#include <stdbool.h>
#include "spinor.h"

int main(void) {
    sp_print(user_print(sp_make_str("Hello, World!")));
    return 0;
}
```

`user_print` という関数はどこにも定義されていないため、リンクどころかコンパイル段階で失敗する。`runtime/spinor.h` には `sp_print(SpObject* obj)` のみが宣言されている。

## 期待される動作 (Expected)

`(print "Hello, World!")` は次のいずれかにコンパイルされるべき:

- `sp_print(sp_make_str("Hello, World!"));` — Spinor の `print` プリミティブを `runtime/spinor.h` の `sp_print` に直接マッピング
- もしくは外側 `sp_print` の二重ラップを取り除き、内側の値変換だけ走らせる

## 影響範囲 (Impact)

- 通常モードの `spinor build` (Issue #47 完了後の `--emit-c` 無し経路) が AOT 出力を gcc でビルドできない。
- Issue #44 (Unikernel 化) の Phase 1 PoC で、`--emit-c` で取り出した C コードをそのまま Unikraft に流し込んだ場合も同じエラーで詰まる可能性が高い (gcc レベルのバグなので Unikraft の musl libc 環境でも同様に失敗する)。
- Issue #47 の `--emit-c` フラグ自体は **正しく動作している** (PR #49)。本バグは codegen 側の独立した問題。

## 原因の所在 (Suspected Location)

`src/Spinor/Compiler/Codegen.hs` の `mangle` 関数 (および `compileExpr` 内で関数呼び出しを処理する箇所)。

- `mangle` がすべての Spinor シンボルに無条件で `user_` プレフィクスを付与しているため、`print` のような **runtime/spinor.h で定義済みのプリミティブ** に対しても `user_print` を生成してしまう。
- `Spinor.Primitive` モジュール (Haskell 評価器側のプリミティブ束縛) は `print` を扱えるが、AOT codegen 側はそれを認識していない。

## 修正方針 (案)

1. `runtime/spinor.h` で公開されているプリミティブシンボルと、対応する Spinor 名のマッピングテーブルを `Codegen.hs` に追加する。
   - 例: `[("print", "sp_print"), ("+", "sp_add"), ("-", "sp_sub"), ("*", "sp_mul"), ("/", "sp_div"), ("=", "sp_eq"), ("<", "sp_lt"), (">", "sp_gt"), ("string-append", "sp_str_append"), ...]`
2. 関数呼び出しの compile 経路で、シンボルがプリミティブテーブルに該当するなら `mangle` を経由せず直接 `sp_*` 名を使う。該当しない場合のみ従来通り `mangle` する。
3. (代替案) `mangle` 関数自体にプリミティブ判定を内包させ、プリミティブの場合は `sp_*` を返すようにする。

## 完了条件 (Definition of Done)

- [ ] `echo '(print "Hello, World!")' > hello.spin && cabal run spinor -- build hello.spin && ./hello` が `Hello, World!` を出力する。
- [ ] 既存の `fib-aot.spin`、`test-tco.spin` などの AOT サンプルが引き続きビルド・実行できる (regression 無し)。
- [ ] `cabal run spinor -- build --emit-c hello.spin` で生成された `hello.c` を `gcc -Iruntime hello.c runtime/spinor.c -o hello` で手動ビルドしても通る。
- [ ] `runtime/spinor.h` の他のプリミティブ (`sp_add`, `sp_str_append` 等) を使う Spinor コードも同様にビルドできる。
- [ ] `Spinor.Primitive` の名前と AOT codegen 側の名前のマッピングが文書化される (コメントまたは specs/)。

## 関連 Issue

- #47 (CLI: `--emit-c` フラグの追加 — 本バグを露見させた発見元)
- #44 (Unikernel 化: AOT 出力の C ツールチェイン互換性が前提)

## 補足: 発見の経緯

PR #49 (Issue #47) の実機検証中、`hello.spin → cabal run spinor -- build hello.spin` で gcc コンパイルエラーが発覚。本バグは PR #49 の変更前から master に存在しており (`git diff origin/master src/Spinor/Compiler/Codegen.hs` は空)、`--emit-c` の有無に関係なく発生することを確認済み。

--- a/examples/hello.spin
+++ b/examples/hello.spin
@@ -1,0 +1,9 @@
+; examples/hello.spin - Smoke test for the AOT codegen primitive fix (Issue #50)
+;
+; Usage:
+;   cabal run spinor -- build --emit-c examples/hello.spin   # Cコードのみ
+;   cabal run spinor -- build examples/hello.spin            # ネイティブビルド
+;   ./hello                                                  # 実行
+;
+; Expected output: Hello, World!
+(print "Hello, World!")

--- a/runtime/spinor.c
+++ b/runtime/spinor.c
@@ -224,10 +224,10 @@ SpObject* sp_file_exists(SpObject* path) {
 
 /* ========== ユーティリティ ========== */
 
-void sp_print(SpObject* obj) {
+SpObject* sp_print(SpObject* obj) {
     if (!obj) {
         printf("NULL\n");
-        return;
+        return obj;
     }
     switch (obj->type) {
         case SP_NIL:
@@ -246,4 +246,5 @@ void sp_print(SpObject* obj) {
             printf("<unknown>\n");
             break;
     }
+    return obj;  /* Lisp 伝統: print は引数をそのまま返す */
 }

--- a/runtime/spinor.h
+++ b/runtime/spinor.h
@@ -84,7 +84,9 @@ SpObject* sp_file_exists(SpObject* path);
 
 /* ========== ユーティリティ ========== */
 
-void sp_print(SpObject* obj);
+/* sp_print: 値を表示し、引数 obj をそのまま返す (Lisp 伝統に倣う)。
+ * 戻り値を SpObject* にしたことで、(print x) を式コンテキストでも使える。 */
+SpObject* sp_print(SpObject* obj);
 const char* sp_format(SpObject* obj);
 
 /* ========== 旧 API との互換性 (エイリアス) ========== */

--- a/src/Spinor/Compiler/Codegen.hs
+++ b/src/Spinor/Compiler/Codegen.hs
@@ -332,7 +332,13 @@ mangle name = "user_" <> T.map sanitize name
 -- | トップレベル式を C のステートメントに変換
 --
 -- 式を評価し、その結果を sp_print で表示する。
+-- ただし `(print x)` のような副作用プリミティブが直接トップレベルにある場合は
+-- 二重出力を避けるため、外側ラップを省略してプリミティブ呼び出しのみ生成する。
 compileStmt :: Expr -> CCode
+-- `(print x)` を top-level に書いた場合: 外側の sp_print 自動ラップをスキップ
+-- (compileExpr が `sp_print(...)` を生成するため、二重出力を防ぐ)
+compileStmt expr@(EList _ [ESym _ "print", _]) =
+    "    " <> compileExpr expr <> ";"
 compileStmt expr =
     let valCode = compileExpr expr
     in "    sp_print(" <> valCode <> ");"
@@ -393,6 +399,13 @@ compileExpr (EList _ [ESym _ "append-file", path, content]) =
 compileExpr (EList _ [ESym _ "file-exists?", path]) =
     "sp_file_exists(" <> compileExpr path <> ")"
 
+-- 出力プリミティブ
+-- `print` は runtime/spinor.h の sp_print に直接マッピングする (mangle 経由しない)。
+-- sp_print は SpObject* を返す (Lisp 伝統: print は引数をそのまま返す) ので
+-- 式コンテキストでも安全に使える。
+compileExpr (EList _ [ESym _ "print", x]) =
+    "sp_print(" <> compileExpr x <> ")"
+
 -- OpenGL プリミティブ
 compileExpr (EList _ [ESym _ "gl-init", w, h, title]) =
     "sp_gl_init(" <> compileExpr w <> ", " <> compileExpr h <> ", " <> compileExpr title <> ")"
@@ -430,6 +443,7 @@ compileExpr (EList _ (ESym _ fname : args))
         in cFun <> "(" <> cArgs <> ")"
   where
     primitives = ["+", "-", "*", "/", "=", "<", ">", "<=", ">=", "if", "defun",
+                  "print",
                   "string-append", "string-length", "substring", "string=?",
                   "read-file", "write-file", "append-file", "file-exists?",
                   "gl-init", "gl-clear", "gl-draw-points",

